pyfixest/estimation/feiv_.py

```diff
@@ -193,3 +206,263 @@ def get_fit(self) -> None:
         # Compute bread matrix
         D = np.linalg.inv(self._tXZ @ self._tZZinv @ self._tZX)
         self._bread = H.T @ D @ H
+
+    def first_stage(self) -> None:
+        """Implement First stage regression."""
+        from pyfixest.estimation.estimation import feols
+
+        #  Start First Stage
+        _X = self._X
+        _Z = self._Z
+        _Y = self._Y
+
+        # Store names of instruments from Z matrix
+        self._non_exo_instruments = list(set(self._coefnames_z) - set(self._coefnames))
+
+        # Prepare your data
+        # Select the columns from self._data that match the variable names
+        data = pd.DataFrame(self._Z_1st_stage, columns=self._coefnames_z)
+
+        # Store instrument variable matrix for future use
+        _Z_iv = data[self._non_exo_instruments]
+        self._Z_iv = _Z_iv.to_numpy()
+        # Dynamically create the formula string
+        # Extract names of fixed effect
+
+        data["Y"] = self._endogvar_1st_stage
+        independent_vars = " + ".join(
+            data.columns[:-1]
+        )  # All columns except the last one ('Y')
+
+        # Set fix effects, cluster, and weight options to be passed to feols.
+
+        if self._has_fixef:
+            FE_vars = self._fixef.split("+")
+            data[FE_vars] = self._data[FE_vars]
+            formula = f"Y ~ {independent_vars} | {self._fixef}"
+        else:
+            formula = f"Y ~ {independent_vars}"
+
```

Comment:
Here, an alternative option would be to add full `FixestFormula` to the `Feiv` model via the [FixestMulti.add_fixest_multi_context](https://github.com/py-econometrics/pyfixest/blob/d77a7efee9397c860303ede8f44d83bad15a52a7/pyfixest/estimation/FixestMulti_.py#L395C21-L395C45) method. 

Then we could obtain the first stage as stored in the `FixestFormula` and as used in [model_matrix_fixest](https://github.com/py-econometrics/pyfixest/blob/d77a7efee9397c860303ede8f44d83bad15a52a7/pyfixest/estimation/model_matrix_fixest_.py#L74). 

I'm open to both the current and the FixestFormula solution, with slight preference for the FixestFormula =) What do you think?

Reply:
Please, correct me if I didn't understand this well. Originally, I directly initiated Feols class to do the first stage like the following; 

```python
        #  Start First Stage

        model1 = Feols(
            Y=_endogvar,
            X=_Z,
            weights=np.ones(self._weights.shape[0]).reshape(-1, 1),
            collin_tol=self._collin_tol,
            coefnames=self._coefnames_z,
            weights_name=None,
            weights_type=None,
        )
        model1.get_fit()
```

The issue that I had is whenever I tried to implement inference such as `wald_test`, I always encountered the issues that the necessary attributes are not defined, and had to define the attributes to proceed. I think [FixestMulti.add_fixest_multi_context](https://github.com/py-econometrics/pyfixest/blob/d77a7efee9397c860303ede8f44d83bad15a52a7/pyfixest/estimation/FixestMulti_.py#L395C21-L395C45) method allows to add more arguments(i.e. context) to define the attributes that might be needed for doing other tasks(such as `wald_test`)? I was actually not too familiar with `FixestFormula` so this will take more time to fix the code, but I would implement this way if this is the better option.

Reply:
Yes, absolutely fine to stick with `feols()` here I think! I'll have to check though what's been going wrong with `Feols` and `add_fixest_multi_context` - in my mind, you should have all the attributed available after calling `get_fit()`,  `vcov()` and `add_fixest_multi_context` on a `Feols` object 🤔 `get_fit()` computes the estimates and residuals, `vcov()` the vcov and adds attributes for small sample corrections & also calls `get_inference()`, `add_fixest_multi_context` adds context outside of the scope of `Feols` (i.e. the model formula, which `Feols` is agnostic of, or the input data frame if specificed by the user).

Reply:
Thanks for the comment. I will look into this!